core/src/main/java/io/questdb/griffin/engine/join/AsOfJoinIndexedRecordCursorFactory.java

```diff
@@ -0,0 +1,204 @@
+        }
+
+        @Override
+        protected void performKeyMatching(long masterTimestamp) {
+            // determine the integer key of the symbol to find in the slave table
+            CharSequence masterSymbolValue = masterRecord.getSymA(masterSymbolColumnIndex);
+            StaticSymbolTable symbolTable = slaveTimeFrameCursor.getSymbolTable(slaveSymbolColumnIndex);
+            int symbolKey = TableUtils.toIndexKey(symbolTable.keyOf(masterSymbolValue));
+            if (symbolKey < 0) {
+                record.hasSlave(false);
+                return;
+            }
+
+            // go backwards through per-frame symbol indexes, until we find a match or exhaust the search space
+            long rowMax = Rows.toLocalRowID(slaveRecB.getRowId());
+            int frameIndex = slaveTimeFrame.getFrameIndex();
+            for (; ; ) {
+                BitmapIndexReader indexReader = slaveTimeFrameCursor.getIndexReaderForCurrentFrame(
+                        slaveSymbolColumnIndex,
+                        BitmapIndexReader.DIR_BACKWARD
+                );
+                // indexReader.getCursor() takes absolute row IDs, but TimeFrameCursor uses numbering relative to
+                // the first row within the BETWEEN ... AND ... range selected by the query.
+                // Use Record.getUpdateRowId() to get the absolute row ID.
+                slaveTimeFrameCursor.recordAt(slaveRecA, Rows.toRowID(frameIndex, slaveTimeFrame.getRowLo()));
+                final long rowLo = Rows.toLocalRowID(slaveRecA.getUpdateRowId());
+                RowCursor rowCursor = indexReader.getCursor(false, symbolKey, rowLo, rowMax + rowLo);
+
+                // Check the first entry only. They are sorted descending by timestamp,
+                // so there aren't any entries more recent than the first one.
+                if (rowCursor.hasNext()) {
+                    long rowId = rowCursor.next();
+                    slaveTimeFrameCursor.recordAt(slaveRecB, Rows.toRowID(frameIndex, rowId));
+                    long slaveTimestamp = scaleTimestamp(slaveRecB.getTimestamp(slaveTimestampIndex), slaveTimestampScale);
+                    if (slaveTimestamp <= masterTimestamp) {
+                        // Enforce tolerance limit if specified
+                        boolean hasSlave = toleranceInterval == Numbers.LONG_NULL ||
+                                slaveTimestamp >= masterTimestamp - toleranceInterval;
+                        record.hasSlave(hasSlave);
+                        return;
+                    }
+                }
+
+                // no match in this frame, try the previous frame
+                if (!slaveTimeFrameCursor.prev()) {
+                    record.hasSlave(false);
+                    return;
+                }
+                slaveTimeFrameCursor.open();
+                frameIndex = slaveTimeFrame.getFrameIndex();
+                rowMax = slaveTimeFrame.getRowHi() - 1;
```

Comment:
We need to place `circuitBreaker.statefulThrowExceptionIfTripped();` somewhere here to avoid leaking readers and worker threads in case of very slow query.